File: 백준/Gold/1197. 최소 스패닝 트리/최소 스패닝 트리.py
import sys
read = sys.stdin.readline

v, e = map(int, read().split())
edges = []
nodes = [i for i in range(v + 1)]

for i in range(e):
    a, b, c = map(int, read().split())
    edges.append((c, a, b))

edges.sort()

def find(a):
    if a == nodes[a]:
        return a
    else:
        nodes[a] = find(nodes[a])
        return nodes[a]

def union(a, b):
    a = find(a)
    b = find(b)
    if a != b:
        nodes[b] = a

weights = []
for weight, start, end in edges:
    if len(weights) == v - 1:
        break
    if find(start) != find(end):
        union(start, end)
        weights.append(weight)

print(sum(weights))

File: 백준/Gold/1197. 최소 스패닝 트리/최소 스패닝 트리.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.PriorityQueue;
import java.util.StringTokenizer;

public class Main {
    static class Edge implements Comparable<Edge> {
        int nodeA, nodeB, weight;

        public Edge (int nodeA, int nodeB, int weight) {
            this.nodeA = nodeA;
            this.nodeB = nodeB;
            this.weight = weight;
        }

        @Override
        public int compareTo(Edge o) {
            return this.weight - o.weight;
        }
    }

    public static int[] parent;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        int V = Integer.parseInt(st.nextToken());
        int E = Integer.parseInt(st.nextToken());

        PriorityQueue<Edge> edges = new PriorityQueue<>();

        for (int i = 0; i < E; i++) {
            st = new StringTokenizer(br.readLine());

            int A = Integer.parseInt(st.nextToken());
            int B = Integer.parseInt(st.nextToken());
            int C = Integer.parseInt(st.nextToken());

            edges.offer(new Edge(A, B, C));
        }

        parent = new int[V + 1];

        for (int i = 1; i <= V; i++) {
            parent[i] = i;
        }

        int minWeight = 0;
        int count = 0;

        while (!edges.isEmpty()) {
            Edge edge = edges.poll();

            if (union(edge.nodeA, edge.nodeB)) {
                minWeight += edge.weight;
                count++;
            }

            if (count == V - 1) {
                break;
            }
        }

        System.out.println(minWeight);
    }

    public static boolean union (int a, int b) {
        int parentA = find(a);
        int parentB = find(b);

        if (parentA != parentB) {
            parent[parentB] = parentA;

            return true;
        }

        return false;
    }

    public static int find (int a) {
        if (a == parent[a]) {
            return a;
        }

        return parent[a] = find(parent[a]);
    }
}